src/main/java/com/aws/greengrass/device/ClientDevicesAuthIPCAgent.java

```diff
@@ -0,0 +1,212 @@
+
+            try {
+                certificateUpdate.setEncodedCertificate(certificate.getEncoded());
+            } catch (CertificateEncodingException e) {
+                logger.atError().log("Unable to encode generated certificate");
+                throw new ServiceError(String.format("Subscribe to certificate update failed with error %s", e));
+            }
+            CertificateUpdateEvent event = new CertificateUpdateEvent();
+            event.setCertificateUpdate(certificateUpdate);
+            this.sendStreamEvent(event);
+        }
+
+        private void forwardClientCertificates(X509Certificate... chain) {
+            // TODO: Update SDK model for accepting certificate chain?
+            CertificateUpdate certificateUpdate = new CertificateUpdate();
+
+            try {
+                certificateUpdate.setEncodedCertificate(chain[0].getEncoded());
+            } catch (CertificateEncodingException e) {
+                logger.atError().log("Unable to encode generated certificate");
+                throw new ServiceError(String.format("Subscribe to certificate update failed with error %s", e));
+            }
+            CertificateUpdateEvent event = new CertificateUpdateEvent();
+            event.setCertificateUpdate(certificateUpdate);
+            this.sendStreamEvent(event);
+        }
+
+        private String getCSR(byte[] publicKey, byte[] privateKey, String algorithm)
+                throws IOException, OperatorCreationException {
+            KeyPair keyPair = new KeyPair(getPublicKeyFromString(publicKey, algorithm),
+                    getPrivateKeyFromString(privateKey, algorithm));
+            return CertificateRequestGenerator
+                    .createCSR(keyPair, serviceName, getHostIpAddresses(), Collections.singletonList("localhost"));
+        }
+
+        private PublicKey getPublicKeyFromString(byte[] encodedPublicKey, String algorithm) {
+            try {
+                KeyFactory factory = KeyFactory.getInstance(algorithm);
+                return factory.generatePublic(new X509EncodedKeySpec(encodedPublicKey));
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                logger.atError("Unable to generate public key from the given encoded string");
+            }
+            return null;
+        }
+
+        private PrivateKey getPrivateKeyFromString(byte[] encodedPrivateKey, String algorithm) {
+            try {
+                KeyFactory factory = KeyFactory.getInstance(algorithm);
+                return factory.generatePrivate(new PKCS8EncodedKeySpec(encodedPrivateKey));
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                logger.atError("Unable to generate private key from the given encoded string");
+            }
+            return null;
+        }
+
+        private List<InetAddress> getHostIpAddresses() throws SocketException {
```

Comment:
This method isn't needed. IP information isn't taken from the CSR, you can just include an empty list